src/metabase/automagic_dashboards/core.clj

```diff
@@ -44,23 +44,32 @@
 (def ^:private ^{:arglists '([field])} id-or-name
   (some-fn :id :name))
 
-(s/defn ->field
+(s/defn ->field :- (s/maybe (type Field))
   "Return `Field` instance for a given ID or name in the context of root."
-  [root id-or-name :- (s/cond-pre su/IntGreaterThanZero su/NonBlankString mbql.s/Field)]
-  (let [id-or-name (if (sequential? id-or-name)
-                     (filters/field-reference->id id-or-name)
-                     id-or-name)]
-    (if (->> root :source (instance? (type Table)))
-      (Field id-or-name)
-      (when-let [field (->> root
-                            :source
-                            :result_metadata
-                            (m/find-first (comp #{id-or-name} :name)))]
-        (-> field
-            (update :base_type keyword)
-            (update :semantic_type keyword)
-            field/map->FieldInstance
-            (classify/run-classifiers {}))))))
+  [{{result-metadata :result_metadata} :source, :as root}
+   field-id-or-name-or-clause :- (s/cond-pre su/IntGreaterThanZero su/NonBlankString mbql.s/Field)]
+  (let [id-or-name (if (sequential? field-id-or-name-or-clause)
+                     (filters/field-reference->id field-id-or-name-or-clause)
+                     field-id-or-name-or-clause)]
+    (or
+     ;; Handle integer Field IDs.
+     (when (integer? id-or-name)
+       (Field id-or-name))
+     ;; handle field string names. Only if we have result metadata. (Not sure why)
+     (when (string? id-or-name)
+       (when-not result-metadata
+         (log/warn (trs "Warning: Automagic analysis context is missing result metadata. Unable to resolve Fields by name.")))
+       (when-let [field (m/find-first #(= (:name %) id-or-name)
+                                      result-metadata)]
+         (-> field
+             (update :base_type keyword)
+             (update :semantic_type keyword)
+             field/map->FieldInstance
+             (classify/run-classifiers {}))))
+     ;; otherwise this isn't returning something, and that's probably an error. Log it.
+     (log/warn (str (trs "Cannot resolve Field {0} in automagic analysis context" field-id-or-name-or-clause)
+                    \newline
+                    (u/pprint-to-str root))))))
```

Comment:
One major bug I was running into was that `->field` would return `nil` in some cases -- I'm not really 100% sure what the original intention was here but I tried to comment it a bit and log a warning if this is returning `nil`

Reply:
I think this was an interaction with the big ol' field array change in MBQL that got in months ago, which never was reconcilliated with this xray codebase